### ISSUE TITLE
[tabular] Improve HPO Documentation and Logic

### ISSUE THREAD

**Issue by AnirudhDagar** [autogluon/autogluon#4075]
**Bug Report Checklist**

- [x] I provided code that demonstrates a minimal reproducible example.  
- [x] I confirmed bug exists on the latest mainline of AutoGluon via source install.  

**Describe the bug**

```
DeprecationWarning: ('`chdir_to_trial_dir` is deprecated. Use the RAY_CHDIR_TO_TRIAL_DIR environment variable instead. Set it to 0 to disable the default behavior of changing the working directory.', <class 'DeprecationWarning'>)
```

Ray Tune [deprecated `chdir_to_trial_dir` in v2.8](https://github.com/ray-project/ray/commit/8b337ad74082305e59cb6865c32939bc2cb27fd5), slated to be removed completely in ray v2.11, in favour of the environment variable `RAY_CHDIR_TO_TRIAL_DIR`. We still pass the deprecated argument in our codebase, specifically at the following two places, which raises a deprecation warning as an exception:

https://github.com/autogluon/autogluon/blob/607f112161b688d324fbb3d67ff5ef4c791520b1/core/src/autogluon/core/models/abstract/abstract_model.py#L1541

https://github.com/autogluon/autogluon/blob/607f112161b688d324fbb3d67ff5ef4c791520b1/core/src/autogluon/core/models/ensemble/bagged_ensemble_model.py#L1393

Whenever hyperparam tuning is invoked, autogluon hpo breaks "without actually erroring out". This can be acheived by setting up `hyperparameter_tune_kwargs='auto'` or `hyperparameter_tune_kwargs='local_random'`. See the reproduction script (thanks @ddelange) below for more details.

The interesting fact while diving deep into it was that, I think there is a bug within the bug. This is probably also the reason why this bug/issue sneaked in through our CI. To explain in more detail, when the `DeprecationWarning` is raised, `ray` raises that as a hard exception which should ideally interrupt the python console with an exit code 1. Unfortunately this doesn't happen because we are wrapping the call to the function `hyperparameter_tune` inside a try except block, capturing the exception error and logging that as a warning.

https://github.com/autogluon/autogluon/blob/607f112161b688d324fbb3d67ff5ef4c791520b1/core/src/autogluon/core/trainer/abstract_trainer.py#L2216-L2250

**Expected resolution**

1. Avoid using deprecated argument `chdir_to_trial_dir`
2. I feel this logic for try except block needs to be refactored, an error should be raised in this case. Right now even if I pass something completely random like `hyperparameter_tune_kwargs='random_string_i_madeup'`, even this will not break the program execution flow; a descriptive error should be raised if a random string is passed or if there is an actual error raised, it should be raised instead of being burried under except.
3. `hyperparameter_tune_kwarg` arg should explicitly mention the options in the documentation (right now it seems very [vague](https://auto.gluon.ai/stable/api/autogluon.tabular.TabularPredictor.fit.html#autogluon.tabular.TabularPredictor.fit) to me with the options 'auto' and 'random', while the codebase can accept multiple other keys as shown below. Either the codebase should be cleaned up to be in line with docs or vice-versa. https://github.com/autogluon/autogluon/blob/607f112161b688d324fbb3d67ff5ef4c791520b1/core/src/autogluon/core/hpo/executors.py#L304-L309
4. Add more tests for making sure such errors are captured in CI and not sneak pass our attention in the future.

**To Reproduce**

```
from autogluon.tabular import TabularPredictor
import pandas as pd
import numpy as np

df = pd.DataFrame(np.random.randint(0, 2, size=(100, 100)))
df.columns = [str(col) for col in df.columns]

TabularPredictor(
    verbosity=3, 
    label='0', 
    path='/tmp/ag',
).fit(
    df, 
    time_limit=100, 
    hyperparameters={"NN_TORCH": {}}, 
    hyperparameter_tune_kwargs='auto',
)
```

<details>
<summary>Click to expand output of the above code snippet</summary>

Search for `DeprecationWarning` in the log output below to get to the error

```
Warning: path already exists! This predictor may overwrite an existing predictor! path="/tmp/ag"
No presets specified! To achieve strong results with AutoGluon, it is recommended to use the available presets.
	Recommended Presets (For more details refer to https://auto.gluon.ai/stable/tutorials/tabular/tabular-essentials.html#presets):
	presets='best_quality'   : Maximize accuracy. Default time_limit=3600.
	presets='high_quality'   : Strong accuracy with fast inference speed. Default time_limit=3600.
	presets='good_quality'   : Good accuracy with very fast inference speed. Default time_limit=3600.
	presets='medium_quality' : Fast training time, ideal for initial prototyping.
============ fit kwarg info ============
User Specified kwargs:
{'hyperparameter_tune_kwargs': 'auto'}
Full kwargs:
{'_feature_generator_kwargs': None,
 '_save_bag_folds': None,
 'ag_args': None,
 'ag_args_ensemble': None,
 'ag_args_fit': None,
 'auto_stack': False,
 'calibrate': 'auto',
 'ds_args': {'clean_up_fits': True,
             'detection_time_frac': 0.25,
             'holdout_data': None,
             'holdout_frac': 0.1111111111111111,
             'memory_safe_fits': True,
             'n_folds': 2,
             'n_repeats': 1,
             'validation_procedure': 'holdout'},
 'excluded_model_types': None,
 'feature_generator': 'auto',
 'feature_prune_kwargs': None,
 'holdout_frac': None,
 'hyperparameter_tune_kwargs': 'auto',
 'included_model_types': None,
 'keep_only_best': False,
 'name_suffix': None,
 'num_bag_folds': None,
 'num_bag_sets': None,
 'num_stack_levels': None,
 'pseudo_data': None,
 'refit_full': False,
 'save_bag_folds': None,
 'save_space': False,
 'set_best_to_refit_full': False,
 'unlabeled_data': None,
 'use_bag_holdout': False,
 'verbosity': 3}
========================================
Warning: hyperparameter tuning is currently experimental and may cause the process to hang.
Saving /tmp/ag/learner.pkl
Saving /tmp/ag/predictor.pkl
Beginning AutoGluon training ... Time limit = 100s
AutoGluon will save models to "/tmp/ag"
=================== System Info ===================
AutoGluon Version:  1.1.0b20240409
Python Version:     3.10.0
Operating System:   Darwin
Platform Machine:   arm64
Platform Version:   Darwin Kernel Version 23.3.0: Wed Dec 20 21:30:44 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T6000
CPU Count:          10
GPU Count:          0
Memory Avail:       8.53 GB / 32.00 GB (26.7%)
Disk Space Avail:   7.22 GB / 460.43 GB (1.6%)
	WARNING: Available disk space is low and there is a risk that AutoGluon will run out of disk during fit, causing an exception.
	We recommend a minimum available disk space of 10 GB, and large datasets may require more.
===================================================
Train Data Rows:    100
Train Data Columns: 99
Label Column:       0
AutoGluon infers your prediction problem is: 'binary' (because only two unique label-values observed).
	2 unique label values:  [1, 0]
	If 'binary' is not the correct problem_type, please manually specify the problem_type parameter during predictor init (You may specify problem_type as one of: ['binary', 'multiclass', 'regression'])
Problem Type:       binary
Preprocessing data ...
Selected class <--> label mapping:  class 1 = 1, class 0 = 0
Using Feature Generators to preprocess the data ...
Fitting AutoMLPipelineFeatureGenerator...
	Available Memory:                    8733.75 MB
	Train Data (Original)  Memory Usage: 0.08 MB (0.0% of available memory)
	Inferring data type of each feature based on column values. Set feature_metadata_in to manually specify special dtypes of the features.
	Stage 1 Generators:
		Fitting AsTypeFeatureGenerator...
			Note: Converting 99 features to boolean dtype as they only contain 2 unique values.
			Original Features (exact raw dtype, raw dtype):
				('int64', 'int') : 99 | ['1', '2', '3', '4', '5', ...]
			Types of features in original data (raw dtype, special dtypes):
				('int', []) : 99 | ['1', '2', '3', '4', '5', ...]
			Types of features in processed data (raw dtype, special dtypes):
				('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
			0.1s = Fit runtime
			99 features in original data used to generate 99 features in processed data.
	Stage 2 Generators:
		Fitting FillNaFeatureGenerator...
			Types of features in original data (raw dtype, special dtypes):
				('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
			Types of features in processed data (raw dtype, special dtypes):
				('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
			0.0s = Fit runtime
			99 features in original data used to generate 99 features in processed data.
	Stage 3 Generators:
		Fitting IdentityFeatureGenerator...
			Types of features in original data (raw dtype, special dtypes):
				('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
			Types of features in processed data (raw dtype, special dtypes):
				('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
			0.0s = Fit runtime
			99 features in original data used to generate 99 features in processed data.
		Skipping CategoryFeatureGenerator: No input feature with required dtypes.
		Skipping DatetimeFeatureGenerator: No input feature with required dtypes.
		Skipping TextSpecialFeatureGenerator: No input feature with required dtypes.
		Skipping TextNgramFeatureGenerator: No input feature with required dtypes.
		Skipping IdentityFeatureGenerator: No input feature with required dtypes.
		Skipping IsNanFeatureGenerator: No input feature with required dtypes.
	Stage 4 Generators:
		Fitting DropUniqueFeatureGenerator...
			Types of features in original data (raw dtype, special dtypes):
				('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
			Types of features in processed data (raw dtype, special dtypes):
				('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
			0.0s = Fit runtime
			99 features in original data used to generate 99 features in processed data.
	Stage 5 Generators:
		Fitting DropDuplicatesFeatureGenerator...
			Types of features in original data (raw dtype, special dtypes):
				('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
			Types of features in processed data (raw dtype, special dtypes):
				('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
			0.0s = Fit runtime
			99 features in original data used to generate 99 features in processed data.
	Types of features in original data (exact raw dtype, raw dtype):
		('int64', 'int') : 99 | ['1', '2', '3', '4', '5', ...]
	Types of features in original data (raw dtype, special dtypes):
		('int', []) : 99 | ['1', '2', '3', '4', '5', ...]
	Types of features in processed data (exact raw dtype, raw dtype):
		('int8', 'int') : 99 | ['1', '2', '3', '4', '5', ...]
	Types of features in processed data (raw dtype, special dtypes):
		('int', ['bool']) : 99 | ['1', '2', '3', '4', '5', ...]
	0.2s = Fit runtime
	99 features in original data used to generate 99 features in processed data.
	Train Data (Processed) Memory Usage: 0.01 MB (0.0% of available memory)
Data preprocessing and feature engineering runtime = 0.17s ...
AutoGluon will gauge predictive performance using evaluation metric: 'accuracy'
	To change this, specify the eval_metric parameter of Predictor()
Saving /tmp/ag/learner.pkl
Automatically generating train/validation split with holdout_frac=0.2, Train Rows: 80, Val Rows: 20
User-specified model hyperparameters to be fit:
{
	'NN_TORCH': {},
}
Saving /tmp/ag/utils/data/X.pkl
Saving /tmp/ag/utils/data/y.pkl
Saving /tmp/ag/utils/data/X_val.pkl
Saving /tmp/ag/utils/data/y_val.pkl
Model configs that will be trained (in order):
	NeuralNetTorch: 	{'ag_args': {'hyperparameter_tune_kwargs': 'auto', 'model_type': <class 'autogluon.tabular.models.tabular_nn.torch.tabular_nn_torch.TabularNeuralNetTorchModel'>, 'priority': 25}}
Fitting 1 L1 models ...
Hyperparameter tuning model: NeuralNetTorch ... Tuning model for up to 89.85s of the 99.83s of remaining time.
	Fitting NeuralNetTorch with 'num_gpus': 0, 'num_cpus': 10
Starting generic AbstractModel hyperparameter tuning for NeuralNetTorch model...
	Hyperparameter search space for NeuralNetTorch:
activation:   Categorical['relu', 'elu']
embedding_size_factor:   Categorical[1.0, 0.5, 1.5, 0.7, 0.6, 0.8, 0.9, 1.1, 1.2, 1.3, 1.4]
dropout_prob:   Categorical[0.1, 0.0, 0.5, 0.2, 0.3, 0.4]
learning_rate:   Real: lower=0.0001, upper=0.03
weight_decay:   Real: lower=1e-12, upper=0.1
proc.embed_min_categories:   Categorical[4, 3, 10, 100, 1000]
proc.impute_strategy:   Categorical['median', 'mean', 'most_frequent']
proc.max_category_levels:   Categorical[100, 10, 20, 200, 300, 400, 500, 1000, 10000]
proc.skew_threshold:   Categorical[0.99, 0.2, 0.3, 0.5, 0.8, 0.9, 0.999, 1.0, 10.0, 100.0]
num_layers:   Categorical[2, 3, 4]
hidden_size:   Categorical[128, 256, 512]
use_batchnorm:   Categorical[False, True]
Warning: Exception caused NeuralNetTorch to fail during hyperparameter tuning... Skipping this model.
Traceback (most recent call last):
  File "/Users/anidagar/Desktop/Work/autogluon/core/src/autogluon/core/trainer/abstract_trainer.py", line 2237, in _train_single_full
    else:
  File "/Users/anidagar/Desktop/Work/autogluon/core/src/autogluon/core/models/abstract/abstract_model.py", line 1474, in hyperparameter_tune
    return self._hyperparameter_tune(hpo_executor=hpo_executor, **kwargs)
  File "/Users/anidagar/Desktop/Work/autogluon/core/src/autogluon/core/models/abstract/abstract_model.py", line 1534, in _hyperparameter_tune
    hpo_executor.execute(
  File "/Users/anidagar/Desktop/Work/autogluon/core/src/autogluon/core/hpo/executors.py", line 409, in execute
    analysis = run(
  File "/Users/anidagar/Desktop/Work/autogluon/core/src/autogluon/core/hpo/ray_hpo.py", line 276, in run
    results = tuner.fit()
  File "/Users/anidagar/miniconda3/envs/autogluon/lib/python3.10/site-packages/ray/tune/tuner.py", line 379, in fit
    return self._local_tuner.fit()
  File "/Users/anidagar/miniconda3/envs/autogluon/lib/python3.10/site-packages/ray/tune/impl/tuner_internal.py", line 477, in fit
    analysis = self._fit_internal(trainable, param_space)
  File "/Users/anidagar/miniconda3/envs/autogluon/lib/python3.10/site-packages/ray/tune/impl/tuner_internal.py", line 596, in _fit_internal
    analysis = run(
  File "/Users/anidagar/miniconda3/envs/autogluon/lib/python3.10/site-packages/ray/tune/tune.py", line 701, in run
    raise DeprecationWarning(
DeprecationWarning: ('`chdir_to_trial_dir` is deprecated. Use the RAY_CHDIR_TO_TRIAL_DIR environment variable instead. Set it to 0 to disable the default behavior of changing the working directory.', <class 'DeprecationWarning'>)
('`chdir_to_trial_dir` is deprecated. Use the RAY_CHDIR_TO_TRIAL_DIR environment variable instead. Set it to 0 to disable the default behavior of changing the working directory.', <class 'DeprecationWarning'>)
Saving /tmp/ag/models/trainer.pkl
No base models to train on, skipping auxiliary stack level 2...
Saving /tmp/ag/models/trainer.pkl
Warning: AutoGluon did not successfully train any models
Saving /tmp/ag/models/trainer.pkl
AutoGluon training complete, total runtime = 0.92s ... Best model: "None"
Loading: /tmp/ag/models/trainer.pkl
Warning: No models found, skipping post_fit logic...
Saving /tmp/ag/models/trainer.pkl
Saving /tmp/ag/learner.pkl
Saving /tmp/ag/predictor.pkl
Saving /tmp/ag/__version__ with contents "1.1.0b20240409"
Saving /tmp/ag/metadata.json
TabularPredictor saved. To load, use: predictor = TabularPredictor.load("/tmp/ag")
```

</details>

**Installed Versions**

<details>

INSTALLED VERSIONS
------------------
date                   : 2024-04-10
time                   : 09:19:45.335924
python                 : 3.10.0.final.0
OS                     : Darwin
OS-release             : 23.3.0
Version                : Darwin Kernel Version 23.3.0: Wed Dec 20 21:30:44 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T6000
machine                : arm64
processor              : arm
num_cores              : 10
cpu_ram_mb             : 32768.0
cuda version           : None
num_gpus               : 0
gpu_ram_mb             : []
avail_disk_size_mb     : 7520

accelerate             : 0.21.0
autogluon              : 1.1.0b20240409
autogluon.eda          : 1.1.0b20240409
autogluon.multimodal   : 1.1.0b20240409
autogluon.timeseries   : 1.1.0b20240409
boto3                  : 1.28.50
defusedxml             : 0.7.1
evaluate               : 0.4.0
gluonts                : 0.14.4
ipython                : 8.12.2
ipywidgets             : 8.1.1
jinja2                 : 3.1.2
joblib                 : 1.3.2
jsonschema             : 4.17.3
lightning              : 2.1.4
matplotlib             : 3.6.3
missingno              : 0.5.2
mlforecast             : 0.10.0
networkx               : 3.1
nlpaug                 : 1.1.11
nltk                   : 3.8.1
nptyping               : 2.4.1
numpy                  : 1.24.4
nvidia-ml-py3          : 7.352.0
omegaconf              : 2.2.3
openmim                : 0.3.9
optimum                : 1.17.1
optimum-intel          : 1.15.2
orjson                 : 3.9.15
pandas                 : 2.1.0
phik                   : 0.12.3
Pillow                 : 10.2.0
PyMuPDF                : 1.21.1
pyod                   : 1.1.3
pytesseract            : 0.3.10
pytorch-metric-learning: 1.7.3
pytorch_lightning      : 2.1.4
requests               : 2.31.0
scikit-image           : 0.19.3
scikit-learn           : 1.3.0
scipy                  : 1.11.2
seaborn                : 0.12.2
seqeval                : 1.2.2
shap                   : 0.44.1
statsforecast          : 1.4.0
statsmodels            : 0.14.0
suod                   : 0.0.9
tensorboard            : 2.14.0
text-unidecode         : 1.3
timm                   : 0.9.7
torch                  : 2.1.2
torchmetrics           : 1.2.1
torchvision            : 0.16.2
tqdm                   : 4.65.2
transformers           : 4.38.2
utilsforecast          : 0.0.10
yellowbrick            : 1.5

</details>

cc @Innixma @ddelange

**Comment by Innixma** [comment]
Thanks for deep diving on this @AnirudhDagar! I think we should keep this either as open or create a separate issue to track the list of improvements you have mentioned after #4076 is merged.

Indeed the documentation for HPO can be improved a lot. Currently it is in a weird state because I don't recommend users to actually do HPO, as it is very likely going to get a worse result than simply setting `best_quality` due to the use of Zeroshot-HPO in `best_quality`.

**Comment by AnirudhDagar** [comment]
Sure, let's keep this open and we can work on the other non-urgent improvements.
